### PR TITLE
Refactor: Server Response Handler

### LINKED ISSUE #648: Use a response handler to convert Response to HttpResponse
**What is not easy to do right now?**
The `Handler.scala` file contains a lot of code to handle `Response` and `HttpData`, which is hard to maintain and refactor.

**Describe the solution you'd like**
Create a new ChannelHandler that can be added at the time of channel initialization, to handle the response types on the Server.

**Why is it important?**
It will make the code inside of `Handler.scala` much easier to refactor.

**Additional context**
Some similar work is done in the client #636

### PR DESCRIPTION
**Pull request by gciuloaica**
- [x]  Need to resolve one memory leak in the server side content handler.

Closes #648

### CODE REVIEW

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 474029.17**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 461922.94**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 481289.95**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 474189.59**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 487956.38**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 444277.21**

File: zio-http/src/main/scala/zhttp/http/HttpData.scala [hunk 1]
```diff
@@ -90,6 +90,9 @@ object HttpData {
   def fromStream[R, E](stream: ZStream[R, E, String], charset: Charset = HTTP_CHARSET): HttpData[R, E] =
     HttpData.BinaryStream(stream.map(str => Unpooled.copiedBuffer(str, charset)))
 
+  def fromByteBufStream[R, E](stream: ZStream[R, E, ByteBuf]): HttpData[R, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpData.scala]:
```suggestion
  private [zhttp] def fromByteBufStream[R, E](stream: ZStream[R, E, ByteBuf]): HttpData[R, E] =
```

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 2]
```diff
@@ -24,17 +20,22 @@ private[zhttp] final case class Handler[R](
     with WebSocketUpgrade[R] { self =>
 
   override def channelRead0(ctx: ChannelHandlerContext, jReq: FullHttpRequest): Unit = {
-    jReq.touch("server.Handler-channelRead0")
     implicit val iCtx: ChannelHandlerContext = ctx
+    val headers                              = Headers.make(jReq.headers())
+    val content                              = Unpooled.copiedBuffer(jReq.content())
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]: Why is this required? It's going to affect performance.

ALSO — If heap is the way to go for us, then we can set that configuration one time when the server starts I think.

File: zio-http/src/main/scala/zhttp/service/Server.scala [hunk 3]
```diff
@@ -39,7 +40,7 @@ sealed trait Server[-R, +E] { self =>
 
 object Server {
   private[zhttp] final case class Config[-R, +E](
-    leakDetectionLevel: LeakDetectionLevel = LeakDetectionLevel.SIMPLE,
+    leakDetectionLevel: LeakDetectionLevel = LeakDetectionLevel.PARANOID,
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Server.scala]: We shouldn't keep `PARANOID` as default. It affects performance quite badly and most people will deploy this on production without changing the configuration.

File: zio-http/src/main/scala/zhttp/service/server/content/handler/ServerResponseHandler.scala [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import zio.stream.ZStream
+import zio.{UIO, ZIO}
+
+/**
+ * Transforms an internal Response into a HttpResponse.
+ */
+@Sharable
+private[zhttp] case class ServerResponseHandler[R](runtime: HttpRuntime[R], serverTime: ServerTimeGenerator)
+    extends SimpleChannelInboundHandler[Either[Throwable, Response[R, Throwable]]](false) {
+  override def channelRead0(ctx: ChannelHandlerContext, response: Either[Throwable, Response[R, Throwable]]): Unit = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/content/handler/ServerResponseHandler.scala]:
```suggestion
  override def channelRead0(ctx: ChannelHandlerContext, response: Response[R, Throwable]): Unit = {
```

Let's handle any kind of Response over here.

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 5]
```diff
@@ -154,30 +118,25 @@ private[zhttp] final case class Handler[R](
...
-              releaseRequest(jReq)
+              unsafeWriteLastContent(data, res)
 
             case data @ HttpData.BinaryChunk(_) =>
-              unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
-              releaseRequest(jReq)
+              unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize), res)
+
+            case HttpData.BinaryStream(_) =>
+              unsafeWriteStreamContent(res)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala:134]: We should move the code to handle `HttpData` also to the ResponseHandler

File: zio-http/src/main/scala/zhttp/service/server/content/handler/ServerResponseHandler.scala [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      }
+    } catch {
+      case err: Throwable => ctx.writeAndFlush(serverErrorResponse(err))
+    }
+    ()
+
+  }
+
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    ctx.writeAndFlush(serverErrorResponse(cause)): Unit
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/content/handler/ServerResponseHandler.scala:54]: needs unit tests. Also this won't really fix #686. Reason being — Once the connection is closed, you can't write anything more on the channel. That's the reason you got this error in the first place.

**Review by tusharmath — CHANGES_REQUESTED** [review]

**Comment by gciuloaica** [comment]
@tusharmath I will close this PR as a did another implementation from scratch today.

**Pull request closed without merging**